### PR TITLE
global: add testlib module for reusability

### LINKED ISSUE #100: Refactor tests code
We should move all the reusable code to inside `hepcrawl`, in a module called `testlib` or similar, specially the **celery tasks** and the **celery monitor**.

That would allow us to remove the `init.py` files and leave the tests directory as pytest likes it, to avoid name collisions if at any time we use tox:

https://docs.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules

### PR DESCRIPTION
**Pull request by spirosdelviniotis**
* Adds celery_monitor module to `testlib`.
* Adds tasks module to `testlib`.
* Adds custom `scrapyd` runner for coverage.
* Adds `fixtures` module instead of `__init__.py` for useful test utils.
* Re-factored `fixtures.py`.

Closes #100

Signed-off-by: Spiros Delviniotis <spyridon.delviniotis@cern.ch>

### CODE REVIEW

File: tests/unit/__init__.py [hunk 1]
```diff
@@ -1 +1,10 @@
- 
+# -*- coding: utf-8 -*-
```
Comment (david-caro) [inline tests/unit/__init__.py]: This file can be removed

Comment (spirosdelviniotis) [inline tests/unit/__init__.py]: We need that file because there are several imports inside unit tests.
E.g.  in the `tests/unit/test_alpha.py`: 
```python
from .responses import fake_response_from_file
```

Comment (david-caro) [inline tests/unit/__init__.py]: Can you move that function to the testlib module too? so we don't need the extra init file?

File: hepcrawl/testlib/tasks.py [hunk 2]
```diff
@@ -11,12 +11,11 @@
 
 from __future__ import absolute_import, print_function, unicode_literals
 
-import json
-
 from six.moves.urllib.parse import urlparse
-
 from celery import Celery
 
+import json
```
Comment (david-caro) [inline hepcrawl/testlib/tasks.py]: revert this change

File: hepcrawl/testlib/celery_monitor.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    @classmethod
+    def do_crawl(cls,
+                 app,
+                 monitor_timeout,
+                 monitor_iter_limit,
+                 crawler_instance,
+                 project='hepcrawl',
+                 spider='WSP',
+                 settings=None,
+                 **crawler_arguments):
```
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: Wasn't this reformatted?

File: hepcrawl/testlib/celery_monitor.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        if settings is None:
+            settings = {}
```
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: Usually, it's a bit better to use:
```python
settings = settings or {}
```

It's a bit faster (the if is not always run), and more concise, while still quite readable.

File: hepcrawl/testlib/celery_monitor.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+        with cls(app, monitor_timeout=monitor_timeout, monitor_iter_limit=monitor_iter_limit) as my_monitor:
+            crawler_instance.schedule(
+                project=project,
+                spider=spider,
+                settings=settings or {},
```
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: Actually, you already did set the value of `settings` before, with this line, the previous is not needed (I think it's better to leave the previous one, and here just use `settings` without `or`).

Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: Because having logic inline with the parameter of a call is quite harder to debug than having the logic by itself.

File: hepcrawl/testlib/fixtures.py [hunk 6]
```diff
@@ -67,4 +69,4 @@ def get_node(spider, tag, response=None, text=None, rtype="xml"):
         selector = Selector(text=text, type=rtype)
     spider._register_namespaces(selector)
     node = selector.xpath(tag)
-    return node
+    return node
```
Comment (david-caro) [inline hepcrawl/testlib/fixtures.py]: missing newline

File: hepcrawl/testlib/tasks.py [hunk 7]
```diff
@@ -14,7 +14,6 @@
 import json
 
 from six.moves.urllib.parse import urlparse
-
```
Comment (david-caro) [inline hepcrawl/testlib/tasks.py]: This space was actually ok:

```
stdlib imports

3rd party imports

current package imports
```

File: tests/unit/test_alpha.py [hunk 8]
```diff
@@ -10,17 +10,23 @@
 from __future__ import absolute_import, print_function, unicode_literals
 
 import pytest
+import os
```
Comment (david-caro) [inline tests/unit/test_alpha.py]: this is not needed right? (there's no new line using the `os` module)

File: tests/unit/test_alpha.py [hunk 9]
```diff
@@ -10,17 +10,23 @@
...
 
 
 @pytest.fixture
 def results():
     """Return results generator from the Alpha spider."""
     spider = alpha_spider.AlphaSpider()
-    records = list(spider.parse(fake_response_from_file('alpha/test_1.htm')))
+    records = list(
+        spider.parse(
+            fake_response_from_file('./../../tests/unit/responses/alpha/test_1.htm')
```
Comment (david-caro) [inline tests/unit/test_alpha.py]: Did you forget to add the new module file on the testlib module?

Also, we should handle the path resolution on the function, so here we can just pass a simpler path, like:
```python
fake_response_from_file('alpha/test_1.htm')
```

Comment (david-caro) [inline tests/unit/test_alpha.py]: Just like it was before xd

File: tests/unit/test_dnb.py [hunk 10]
```diff
@@ -54,19 +54,22 @@ def record(scrape_pos_page_body):
     return request.callback(response)
 
 
+@pytest.mark.skip(reason='connecting to external source')
```
Comment (david-caro) [inline tests/unit/test_dnb.py]: Can you add to the message in the skips a reference to the issue to fix it? (create the issue if not there yet please)

File: hepcrawl/testlib/celery_monitor.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        self.monitor_timeout = monitor_timeout
+        self.monitor_iter_limit = monitor_iter_limit
+
+    def __enter__(self):
+        state = self.app.events.State()
+
+        def announce_succeeded_tasks(event):
+            state.event(event)
+            task = state.tasks.get(event['uuid'])
+            print('TASK SUCCEEDED: %s[%s] %s' % (task.name, task.uuid, task.info(),))
```
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: Instead of print, you should be using logging module.

File: hepcrawl/testlib/celery_monitor.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+            tasks = self.app.AsyncResult(task.id)
+            for task in tasks.result:
+                self.results.append(task)
+            self.recv.should_stop = True
+
+        def announce_failed_tasks(event):
+            state.event(event)
+            task = state.tasks.get(event['uuid'])
+            print('TASK FAILED: %s[%s] %s' % (task.name, task.uuid, task.info(),))
```
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]: idem

File: hepcrawl/testlib/celery_monitor.py (hunk 11 above)
Comment (david-caro) [inline hepcrawl/testlib/celery_monitor.py]:
```python
import logging

LOGGER = logger.get_logger(__name__)

class Bla():
    def blo():
        LOGGER.info('My message')
```

File: hepcrawl/testlib/fixtures.py [hunk 13]
```diff
@@ -65,6 +64,12 @@ def get_node(spider, tag, response=None, text=None, rtype="xml"):
...
     elif text:
         selector = Selector(text=text, type=rtype)
+
     spider._register_namespaces(selector)
     node = selector.xpath(tag)
     return node
+
+
+def get_responses_dir_file(*file_names):
+    project_root_dir = os.path.dirname(os.path.abspath(__file__ + os.path.join('.', '..', '..')))
```
Comment (david-caro) [inline hepcrawl/testlib/fixtures.py]:
```python
def get_responses_path(*path_chunks):
"""
    :param path_chunks: Optional extra path element to suffix the responses directory with.
     Example:
         >> get_responses_path()
         '/home/myuser/hepcrawl/tests/responses'

        >> get_responses_path('one', 'two')
        '/home/myuser/hepcrawl/tests/responses/one/two'
"""
os.path.abspath(
    os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        '..',
        '..',
    )
)
```

File: tests/functional/WSP/test_wsp.py [hunk 14]
```diff
@@ -19,82 +19,8 @@
...
-        with cls(app, monitor_timeout=monitor_timeout, monitor_iter_limit=monitor_iter_limit) as my_monitor:
-            crawler_instance.schedule(
-                project=project,
-                spider=spider,
-                settings=settings or {},
-                **crawler_arguments
-            )
-
-        return my_monitor.results
+from hepcrawl.testlib.tasks import app
```
Comment (david-caro) [inline tests/functional/WSP/test_wsp.py]: Instead of using just `app` here better use something like `as celery_app`.

File: hepcrawl/testlib/fixtures.py [hunk 15]
```diff
@@ -65,6 +64,29 @@ def get_node(spider, tag, response=None, text=None, rtype="xml"):
...
+
+def get_responses_path(*path_chunks):
+    """
+    :param path_chunks: Optional extra path element to suffix the responses directory with.
+     Example:
+        >> get_responses_path()
+        '/home/myuser/hepcrawl/tests/responses'
+
+        >> get_responses_path('one', 'two')
+        '/home/myuser/hepcrawl/tests/responses/one/two'
```
Comment (david-caro) [inline hepcrawl/testlib/fixtures.py]: The example should be wrapped in `:` and have a newline right after (I recommend putting it last):
    http://thomas-cokelaer.info/tutorials/sphinx/docstring_python.html
